### DISCUSSION TITLE
firebase-messaging-sw.js:1 Refused to load the script

### DISCUSSION THREAD

**Discussion by irkhamdayat11** [nginx/nginx#257] [Q&A]
i have application deployed using nginx but when i inspect i have error "'https://www.gstatic.com/firebasejs/9.23.0/firebase-app-compat.js' because it violates the following Content Security Policy directive: "default-src 'self' 'unsafe-inline' 'unsafe-eval' googleapis.com gstatic.com jsdelivr.net". Note that 'script-src-elem' was not explicitly set, so 'default-src' is used as a fallback."

this is my nginx file
`upstream svc_testing {
  ip_hash;
  server testing.svc:3344 max_fails=2 fail_timeout=15s;
}

log_format apm '"$time_local" client=$remote_addr '
               'method=$request_method request="$request" '
               'request_length=$request_length '
               'status=$status bytes_sent=$bytes_sent '
               'upstream_addr=$upstream_addr '
               'upstream_status=$upstream_status '
               'request_time=$request_time '
               'upstream_response_time=$upstream_response_time '
               'upstream_connect_time=$upstream_connect_time '
               'upstream_header_time=$upstream_header_time';

server {
  listen 3000;
  server_name localhost;

  add_header Strict-Transport-Security "max-age=31536000; includeSubDomains; preload" always;
  add_header X-Content-Type-Options nosniff;
  add_header X-Frame-Options SAMEORIGIN;
  add_header X-XSS-Protection "1; mode=block";

  # Header Content Security Policy
  add_header Content-Security-Policy "default-src 'self' googleapis.com gstatic.com jsdelivr.net; script-src 'self' 'unsafe-inline' 'unsafe-eval' https://www.gstatic.com https://www.googleapis.com; script-src-elem 'self' https://www.gstatic.com; connect-src 'self' https://fcm.googleapis.com https://www.googleapis.com; worker-src 'self' blob: https://www.gstatic.com https://www.googleapis.com;" always;

  add_header Referrer-Policy "origin-when-cross-origin" always;
  add_header Feature-Policy "camera 'none'; geolocation 'none'; microphone 'none'; payment 'none'; usb 'none'" always;

  server_tokens off;
  add_header Server "Hidden" always;
  proxy_hide_header X-Powered-By;

  error_page 400 403 404 500 502 503 504 /error.html;

  # Konfigurasi halaman error
  location = /error.html {
    internal;
    root /app;
  }

  client_max_body_size 100M;
  root /app;
  index index.html;

  access_log /dev/stdout apm;
  error_log /dev/stderr;

  location /nginx_status {
    stub_status on;
    access_log off;
    allow all;
  }

  location = /robots.txt {
    return 301 http://$host;
  }

  location / {
    add_header Set-Cookie "Path=/; Secure; HttpOnly; SameSite=Strict";
    try_files $uri $uri/ /index.html =404;
  }

  location /firebase-messaging-sw.js {
    root /app/public; 
    default_type "application/javascript";
    add_header Content-Security-Policy "default-src 'self'; script-src 'self' 'unsafe-inline' 'unsafe-eval' https://www.gstatic.com https://www.googleapis.com; worker-src 'self' blob: https://www.gstatic.com;" always; 
    try_files $uri =404;
  }

  location /services/ {
    proxy_cache_revalidate on;
    proxy_cache_min_uses 3;
    proxy_cache_use_stale error timeout updating http_500 http_502 http_503 http_504;
    proxy_cache_background_update on;
    proxy_cache_lock on;

    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header Host $http_host;
    proxy_redirect off;
    proxy_connect_timeout 600s;
    proxy_read_timeout 5400s;
    proxy_send_timeout 5400s;

    proxy_cookie_path / "/; Secure; HttpOnly; SameSite=Strict";

    proxy_pass http://svc_testing/;
  }
}
`

**Comment by Maryna-f5** [comment]
This discussion has been inactive for more than 90 days, closing it.

**Comment by Raphasha27** [comment]
This is likely a MIME type issue — Firebase's `firebase-messaging-sw.js` needs to be served as `text/javascript` or `application/javascript`, but nginx is serving it with the wrong MIME type.

**Fix 1:** Ensure the `.js` extension is mapped correctly:

```nginx
http {
    include       mime.types;  # includes application/javascript for .js
    default_type  application/octet-stream;
}
```

**Fix 2:** If your service worker is at a non-standard path, explicitly set the MIME type:

```nginx
location /firebase-messaging-sw.js {
    alias /path/to/your/static/firebase-messaging-sw.js;
    add_header Content-Type application/javascript;
}
```

**Fix 3:** If behind a reverse proxy, ensure the backend sends the correct header:

```nginx
location /service-worker/ {
    proxy_pass http://frontend:3000;
    proxy_set_header Accept application/javascript;
}
```

**Fix 4:** For CORS issues (also common with service workers):

```nginx
location /firebase-messaging-sw.js {
    add_header Service-Worker-Allowed /;
    add_header Access-Control-Allow-Origin *;
}
```

The error `Refused to load the script` with service workers specifically means the browser rejected the script because:
1. MIME type mismatch (most common — check your Content-Type response header)
2. The script is not served from the scope's origin
3. Missing `Service-Worker-Allowed` header for non-root scopes

Check the response headers:

```bash
curl -I https://yourdomain.com/firebase-messaging-sw.js
```

The `content-type` should be `application/javascript` or `text/javascript`.
